### DISCUSSION TITLE
FASTAPI I18N Internationalization

### DISCUSSION THREAD

**Discussion by maverick42** [fastapi/fastapi#7583] [Questions]
Hi,

I have database translations. I follow this tutorial to implement in fastapi :
https://sqlalchemy-utils.readthedocs.io/en/latest/internationalization.html

my endpoint:

```python
@router.get("/", response_model=List[usrRole])
def read_roles(
    db: Session = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
    current_user: DBUser = Depends(get_current_active_user),
    language: str = Cookie(None),
):
```

my db_model :

```python
# For testing purposes we define this as simple function which returns
# locale 'fi'. Usually you would define this function as something that
# returns the user's current locale.
def get_locale():
    return 'fi' <<< **how to get language cookie here ....**

translation_hybrid = TranslationHybrid(
    current_locale=get_locale,
    default_locale='en'
)

class usrRole(Base):
    __tablename__  = MODULE_NAME + 'role'
    __table_args__ = ({ "schema": DATABASE_SCHEMA})

    role_id           = Column(Integer,   primary_key=True, index=True)
    name_translations = Column(HSTORE)

    # translated column
    name = translation_hybrid(name_translations)
```

All works fine but ...only with 'fi' value :)  How to get language cookie value in the db_model ? 

Regards 
Frédérik

**Comment by retnikt** [comment]
You will need to add a [Cookie](https://fastapi.tiangolo.com/tutorial/cookie-params/) parameter to your `get_current_active_user` dependency, and then pass it to the `get_locale` method somehow. FastAPI is mainly an async framework, but this code appears to have been written for frameworks like Flask that use thread-locals to access request information. This is a question for `sql_alchemy_utils`, but I would recommend you to implement your own i18n, or use a different library.

**Comment by maverick42** [comment]
Thanks retnikt.

However I find this plugin very useful and FASTAPI users could be interested (at least those using postgreSQL).

After a lot of research I found someone :) @https://github.com/omrihar who did a great job of making this plugin work with a middleware ! He gave me his method and in a few minutes I was able to use TranslationHybrid with FASTAPI.

In the same way I could easily communicate the current active user to my audit mixin to update created_by and updated_by table field.

It works perfectly :)

So I can close this issue.

**Comment by tiangolo** [comment]
Thanks for the help here @retnikt ! :clap: :bow: 

Thanks for reporting back and closing the issue @maverick42  :+1:

**Comment by KrunchMuffin** [comment]
> Thanks retnikt.
> 
> However I find this plugin very useful and FASTAPI users could be interested (at least those using postgreSQL).
> 
> After a lot of research I found someone :) @https://github.com/omrihar who did a great job of making this plugin work with a middleware ! He gave me his method and in a few minutes I was able to use TranslationHybrid with FASTAPI.
> 
> In the same way I could easily communicate the current active user to my audit mixin to update created_by and updated_by table field.
> 
> It works perfectly :)
> 
> So I can close this issue.

Be great if you could share with the rest of us. :)

**Comment by rfleschenberg** [comment]
@maverick42 It would be great if you can share some information about the solution you found :) I am currently trying to figure out how to pass a parameter to `get_locale()` and so far, I don't see how that's possible from a middleware.

**Comment by rfleschenberg** [comment]
I just found https://github.com/tomwojcik/starlette-context. This looks like it should work :)

**Comment by julenbadiola** [comment]
My solution: 

```python
from contextvars import ContextVar

from sqlalchemy_utils import TranslationHybrid
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request

POSSIBLE_LOCALES = ["en", "es"]
DEFAULT_LOCALE = "en"

_lang: ContextVar[str] = ContextVar(DEFAULT_LOCALE, default=None)


def get_language() -> str:
    return _lang.get()


translation_hybrid = TranslationHybrid(
    current_locale=get_language,
    default_locale=DEFAULT_LOCALE
)


class RequestContextMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ):
        try:
            lang = request.headers["accept-language"]
            user_language = lang if lang in POSSIBLE_LOCALES else DEFAULT_LOCALE
        except:
            user_language = DEFAULT_LOCALE
        language = _lang.set(user_language)
        response = await call_next(request)
        return response
```

Add the middleware to the fastapi app
```python
app.add_middleware(RequestContextMiddleware)
```

**Reply by Myzel394** [reply]
@maverick42 How do I use this? Should I use `get_language` inside my routes? Doesn't `_lang` make the language global? I'd rather try to get the language on a per request basis
